packages/theme/src/theme.ts

```diff
@@ -65,41 +59,71 @@ export class Theme extends HTMLElement {
     }
 
     get color(): Color {
-        const color = this.getAttribute('color');
-        if (!color) return DefaultColor;
-
-        const colorFragment = Theme.themeFragments.get(color);
-        if (colorFragment && colorFragment.kind === 'color') {
-            return color as Color;
-        }
-        return DefaultColor;
+        return this.getFragementNameByKind('color') as Color;
     }
 
     set color(newValue: Color) {
         this.setAttribute('color', newValue);
     }
 
     get scale(): Scale {
-        const scale = this.getAttribute('scale');
-        if (!scale) return DefaultScale;
-
-        const scaleFragment = Theme.themeFragments.get(scale);
-        if (scaleFragment && scaleFragment.kind === 'scale') {
-            return scale as Scale;
-        }
-        return DefaultScale;
+        return this.getFragementNameByKind('scale') as Scale;
     }
 
     set scale(newValue: Scale) {
         this.setAttribute('scale', newValue);
     }
 
+    private getFragementNameByKind(kind: string): string {
+        const kindFragments = Theme.themeFragmentsByKind.get(
+            kind
+        ) as FragmentMap;
+        /* istanbul ignore if */
+        if (!kindFragments) {
+            throw new Error(`Unknown theme fragment kind '${kind}'`);
+            /* istanbul ignore if */
+        } else if (kindFragments.size === 0) {
+            throw new Error(`No theme fragments of kind '${kind}'`);
+        }
+        const name = this.getAttribute(kind);
+        if (name) {
+            const fragment = kindFragments.get(name);
+            if (fragment) {
+                return name;
+            }
+        }
+        const defaultFragment = kindFragments.get('default');
+        /* istanbul ignore else */
+        if (defaultFragment) {
+            return defaultFragment.name;
+        }
+        throw new Error(
+            `Incorrectly configured theme fragments of kind '${kind}'`
+        );
+    }
+
     private get styles(): CSSResult[] {
-        return [
-            coreStyles,
-            Theme.themeFragment(this.color).styles,
-            Theme.themeFragment(this.scale).styles,
-        ];
+        const themeKinds = [...Theme.themeFragmentsByKind.keys()];
+        const styles = themeKinds.reduce(
+            (acc, kind) => {
+                const kindFragments = Theme.themeFragmentsByKind.get(
+                    kind
+                ) as FragmentMap;
+                const defaultStyles = kindFragments.get('default');
+                // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+                // @ts-ignore
```

Comment:
Is there a way to type this that TS doesn't get angry about?

Reply:
Something like this could work:

```diff
diff --git a/packages/theme/src/theme.ts b/packages/theme/src/theme.ts
index a841e5c1..9a8f67d8 100644
--- a/packages/theme/src/theme.ts
+++ b/packages/theme/src/theme.ts
@@ -47,9 +47,16 @@ export interface ThemeData {
     scale?: Scale;
 }
 
-export class Theme extends HTMLElement {
+type ThemeKindProvider = {
+    [P in FragmentType]: string;
+};
+
+export class Theme extends HTMLElement implements ThemeKindProvider {
     private hasAdoptedStyles = false;
-    private static themeFragmentsByKind: Map<string, FragmentMap> = new Map();
+    private static themeFragmentsByKind: Map<
+        FragmentType,
+        FragmentMap
+    > = new Map();
     private static defaultFragments: Set<FragmentName> = new Set(['core']);
 
     private static templateElement?: HTMLTemplateElement;
@@ -58,6 +65,10 @@ export class Theme extends HTMLElement {
         return ['color', 'scale'];
     }
 
+    get core(): 'core' {
+        return 'core';
+    }
+
     get color(): Color {
         return this.getFragementNameByKind('color') as Color;
     }
@@ -74,7 +85,7 @@ export class Theme extends HTMLElement {
         this.setAttribute('scale', newValue);
     }
 
-    private getFragementNameByKind(kind: string): string {
+    private getFragementNameByKind(kind: FragmentType): string {
         const kindFragments = Theme.themeFragmentsByKind.get(
             kind
         ) as FragmentMap;
@@ -110,8 +121,7 @@ export class Theme extends HTMLElement {
                     kind
                 ) as FragmentMap;
                 const defaultStyles = kindFragments.get('default');
-                // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-                // @ts-ignore
+
                 const { [kind]: name } = this;
                 const currentStyles = kindFragments.get(name);
                 if (currentStyles) {
```

i.e. provide a mapped type to declare that we will have a property for each fragment kind, implement that type, and then change the fragment map to use it.

Reply:
That did the trick with one extra typing in there, thanks! TS is crazy.